src/redisgears_python.c

```diff
@@ -3034,6 +3035,69 @@ static PyObject* modelRunnerRun(PyObject *cls, PyObject *args){
     return tensorList;
 }
 
+static void FinishAsyncModelRun(RAI_OnFinishCtx *onFinishCtx, void *private_data) {
+
+	RAI_Error *error;
+	RedisAI_InitError(&error);
+	RAI_ModelRunCtx* mctx = RedisAI_GetAsModelRunCtx(onFinishCtx, error);
+
+	RedisGearsPy_LOCK
+	PyObject* future = private_data;
+	PyObject* tensorList = PyList_New(0);
+	PyObject* pArgs = PyTuple_New(2);
+	PyTuple_SetItem(pArgs, 0, future);
+
+	if (RedisAI_GetErrorCode(error) != RedisAI_ErrorCode_OK) {
+		const char *errStr = RedisAI_GetError(error);
+		PyErr_SetString(GearsError, errStr);
+		PyTuple_SetItem(pArgs, 1, GearsError);
+		PyObject* r = PyObject_CallObject(setFutureExceptionFunction, pArgs);
+		goto finish;
+	}
+
+	for(size_t i = 0 ; i < RedisAI_ModelRunCtxNumOutputs(mctx) ; ++i){
+		PyTensor* pyt = PyObject_New(PyTensor, &PyTensorType);
+		pyt->t = RedisAI_TensorGetShallowCopy(RedisAI_ModelRunCtxOutputTensor(mctx, i));
+		PyList_Append(tensorList, (PyObject*)pyt);
+		Py_DECREF(pyt);
+	}
+
+	PyTuple_SetItem(pArgs, 1, tensorList);
+	PyObject* r = PyObject_CallObject(setFutureResultsFunction, pArgs);
+
+	finish:
+	Py_DECREF(pArgs);
+	if(!r){
+		char* err = getPyError();
+		RedisModule_Log(NULL, "warning", "Error happened when releasing execution future, error='%s'", err);
+		RG_FREE(err);
+	}else{
+		Py_DECREF(r);
+	}
+	RedisGearsPy_UNLOCK
+}
+
+static PyObject* modelRunnerRunAsync(PyObject *cls, PyObject *args){
+	verifyRedisAILoaded();
+	if(PyTuple_Size(args) != 1){
+		PyErr_SetString(GearsError, "Wrong number of arguments given to modelRunnerRunAsync");
+		return NULL;
+	}
+	PyGraphRunner* pyg = (PyGraphRunner*)PyTuple_GetItem(args, 0);
+	if(!PyObject_IsInstance((PyObject*)pyg, (PyObject*)&PyGraphRunnerType)){
+		PyErr_SetString(GearsError, "Given argument is not of type PyGraphRunner");
+		return NULL;
+	}
+
+	PyObject* pArgs = PyTuple_New(0);
+	PyObject* future = PyObject_CallObject(createFutureFunction, pArgs);
+	Py_DECREF(pArgs);
+
+	RedisAI_ModelRunAsync(pyg->g, FinishAsyncModelRun, future);
```

Comment:
is pyg->g refcounted? if so do we need to increase refcount or does RedisAI_ModelRunAsync will do it? If it's not refcounted then is it ok that it will be freed before the modelrun will finished?